openshift/release
# stackrox: test on dev e2e branch

Instead of the change present in this PR, I have requested:
1. Add additional stackrox metrics to the kube-burner-ocp predefined metrics: https://github.com/kube-burner/kube-burner-ocp/pull/103

---
This PR's latest attempt was:
1. Add a metrics-endpoint arg+file
  - local prometheus endpoint is unknown before run-time (auto-lookup is not done in kube-burner when endpoints are defined).
  - baked-in metrics are not collected (or we'd need to copy the baked-in config from the kube-burner-ocp repo).
  - the prometheus token is required. So we cannot define the local prometheus generically like "prometheus-k8s-openshift-monitoring.apps.${cluster_id}.ocp.infra.rox.systems"
  - when using the metrics-endpoints, the prometheus endpoints are required and auth for it is required.
  - [ ] Add a step layer (chain?) that looks up the local prometheus or creates the metrics-endpoint.yaml file.

Notes:
* Call stack example:
```
job -- ci-operator/config/openshift-qe/ocp-qe-perfscale-ci/openshift-qe-ocp-qe-perfscale-ci-main__aws-4.16-nightly-x86.yaml
-> osci `workflow` * sets up the cluster to run the test with
-> osci `test: chain: openshift-qe-control-plane` -- ci-operator/step-registry/openshift-qe/cluster-density-v2/openshift-qe-cluster-density-v2-chain.yaml 
  -> osci `test: chain: openshift-qe-cluster-density-v2` -- ci-operator/step-registry/openshift-qe/cluster-density-v2/openshift-qe-cluster-density-v2-ref.yaml
    -> runs script -- ci-operator/step-registry/openshift-qe/cluster-density-v2/openshift-qe-cluster-density-v2-commands.sh
      -> exec's `./run.sh` from clone of github.com/cloud-bulldozer/e2e-benchmarking repo -- https://github.com/cloud-bulldozer/e2e-benchmarking/blob/master/workloads/kube-burner-ocp-wrapper/run.sh
        -> runs kube-burner-ocp after downloading from `github.com/kube-burner/kube-burner-ocp/releases/download/v${KUBE_BURNER_VERSION}`
          -> runs internal kube-burner pkg with config based on arguments and bundled yaml configs
```
* kube-burner does not auto-find the local prometheus if a metrics-endpoint is defined: https://github.com/kube-burner/kube-burner-ocp/blob/466287d41157d9c67beed6e7948377895d84cd41/common.go#L55-L56
* kube-burner-ocp bundled configs can be written to disk at runtime, instead of executing the job, and they include the variables before the templates are filled: `kube-burner-ocp ... --extract` (same as defined for example for the cluster-density workload at https://github.com/kube-burner/kube-burner-ocp/blob/main/cmd/config/cluster-density-v2/cluster-density-v2.yml).
   * These can then be modified and then they will be found and used as the config for kube-burner-ocp: https://github.com/davdhacs/e2e-benchmarking/blob/e6ae5c43d54a6f4c561446abb46faa92c5b00e07/workloads/kube-burner-ocp-wrapper/run.sh#L175
* metrics-endpoint use does work -- I was able to run and collect stackrox metrics when defining a metrics-endpoint and setting the correct prometheus endpoint in the yaml for the cluster I was testing.

Failed solutions:
1. Modify the extracted config: https://github.com/davdhacs/e2e-benchmarking/blob/e6ae5c43d54a6f4c561446abb46faa92c5b00e07/workloads/kube-burner-ocp-wrapper/run.sh#L175
  a. declined
1. Add accepting an EXTRA_METRICS_FILE environment variables in kube-burner-ocp: https://github.com/kube-burner/kube-burner-ocp/pull/102
  a. declined
1. Add additional stackrox metrics to the kube-burner-ocp predefined metrics: https://github.com/kube-burner/kube-burner-ocp/pull/103
  a. declined
  b. Re-requesting (see top)
1. metrics-endpoint
    - [x] without prometheus endpoint defined (required)
    - [x] with a localhost:9090 prometheus endpoint defined (not reachable)
    - [x] looking up the local prometheus in the script (this requires changes to the step or e2e-benchmarking scripts)
    - [x] looking up the local prometheus in the EXTRA_FLAGS variable or metrics-endpoint yaml (these are not parsed to eval the commands).

Possible solutions:
1. Fork of the e2e-benchmarking repo OR copies of each step like cluster-density-v2. Then modify the commands or run script to:
  a. lookup the cluster's prometheus endpoint and supply a metrics-endpoint including this.
  b. or modify the config after extracting it
  c. or include a copy of the config files (`--extract` for each job, add stackrox.yaml file and reference in the jobs's yaml).
3. Create a custom release of kube-burner-ocp that includes the stackrox metrics (could be a custom tagged release so that it is not in master or standard runs). Then the KUBE_BURNER_VERSION could be set and the existing call stack will download and use it.
4. Copy all of the openshift-ci workflows, chains, and steps for perf&scale and modify these to include a stackrox metrics.yaml (multiple ways to include it like other attempts/options with kube-burner-ocp's local directory lookup or --extract+sed etc).
5. ?

/hold
/uncc
